packages/block-library/src/table-of-contents/edit.js

```diff
@@ -0,0 +1,134 @@
+/**
+ * Internal dependencies
+ */
+import * as Utils from './utils';
+import ListLevel from './ListLevel';
+
+/**
+ * WordPress dependencies
+ */
+import { Component } from '@wordpress/element';
+import { subscribe } from '@wordpress/data';
+import { __ } from '@wordpress/i18n';
+import {
+	IconButton,
+	Toolbar,
+	PanelBody,
+	ToggleControl,
+} from '@wordpress/components';
+import { BlockControls, InspectorControls } from '@wordpress/editor';
+
+class TOCEdit extends Component {
+	constructor() {
+		super( ...arguments );
+
+		this.state = {
+			wpDataUnsubscribe: null,
+		};
+
+		this.toggleAttribute = this.toggleAttribute.bind( this );
+		this.refresh = this.refresh.bind( this );
+	}
+
+	toggleAttribute( propName ) {
+		const value = this.props.attributes[ propName ];
+		const { setAttributes } = this.props;
+
+		setAttributes( { [ propName ]: ! value } );
+	}
+
+	refresh() {
+		const { setAttributes } = this.props;
+		const headings = Utils.getPageHeadings();
+		setAttributes( { headings } );
+	}
+
+	componentDidMount() {
+		const { attributes, setAttributes } = this.props;
+		const headings = attributes.headings || [];
+		const wpDataUnsubscribe = subscribe( () => {
+			const pageHeadings = Utils.getPageHeadings();
+			this.setState( { pageHeadings } );
+		} );
+
+		setAttributes( { headings } );
```

Comment:
I think I had to keep that there to make the block work when newly created, but I haven't checked again since changing a bunch of other things. I can experiment removing it to see if it still behaves.